src/compiler/transformers/generators.ts

```diff
@@ -164,12 +164,13 @@ namespace ts {
     }
 
     // A generated code block
-    interface CodeBlock {
+    type CodeBlock = | ExceptionBlock | LabeledBlock | SwitchBlock | LoopBlock | WithBlock;
+    interface CodeBlockBase {

```

Comment:
Just move `kind` to each code block with its more specific type and remove this type. That way we can just leverage `kind` as a discriminant.